### PR TITLE
[65311] Transition to Datumaro detection heuristics 

### PR DESCRIPTION
**Pull request by vgoluben**
### All Submissions:
* [x] Create a meaningful name of the PR with reference to JIRA. Example: *[43714] [NIGHTLY] Take correct tag for snyk scans*
* [x] JIRA tickets:
  * 65311 - Transition to Datumaro format rules instead of WB's ones
* [x] Set assignee to the PR.
* [x] Add corresponding labels to the PR. 
* [x] Set appropriate milestone to the PR.
* [x] Added tests for new functionality (unit / e2e)
* [x] Followed code style guidelines:
  * Code covered with typings
  * No debug printing in production code
  * No commented blocks of code in production code

### CODE REVIEW

File: wb/main/jobs/datasets/convert_dataset_job.py [hunk 1]
```diff
@@ -87,14 +87,16 @@ def run_conversion(self, original_dataset: DatasetsModel,
                        dataset_format: DatasetTypesEnum):
         tool = DatumaroTool()
         tool.set_mode(DatumaroModesEnum.convert)
-        tool.set_input_output_paths(original_dataset.path, result_dataset.path)
+        tool.set_path('input-path', original_dataset.path)
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/convert_dataset_job.py:90]: add  the `set_input_path` method

File: wb/main/jobs/datasets/convert_dataset_job.py [hunk 2]
```diff
@@ -87,14 +87,16 @@ def run_conversion(self, original_dataset: DatasetsModel,
...
+        tool.set_path('output-dir', result_dataset.path)
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/convert_dataset_job.py:91]: add  the `set_output_dir` method

File: wb/main/jobs/datasets/convert_dataset_job.py [hunk 3]
```diff
@@ -87,14 +87,16 @@ def run_conversion(self, original_dataset: DatasetsModel,
...
         tool.set_conversion(dataset_format, self._format_conversion_map[dataset_format])
-        tool.enable_image_copy()
+        tool.set_flag('')
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/convert_dataset_job.py:93]: It is absolutely not understandable.  Please create a specific method to add `--`

File: wb/main/jobs/datasets/convert_dataset_job.py [hunk 4]
```diff
@@ -87,14 +87,16 @@ def run_conversion(self, original_dataset: DatasetsModel,
...
+        tool.set_flag('save-images')
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/convert_dataset_job.py:94]: Add a specific method to set this parameter

File: wb/main/shared/enumerates.py [hunk 5]
```diff
@@ -57,5 +59,11 @@ class DatasetTypesEnum(enum.Enum):
     # text
     csv = 'csv'
 
-    def is_nlp(self):
+    @classmethod
+    def get_value(cls, value: str) -> Optional[DatasetTypesEnum]:
```
Comment (artyomtugaryov) [inline wb/main/shared/enumerates.py:63]: It doesn't work:
```
from enum import Enum
class E(Enum):
    e = 'e'
    @classmethod
    def get_value(cls, value: str):
        if E(value):
            return E(value)
        return None
E.get_value('q')
```
gets the error:
```
ValueError: 'q' is not a valid E
```

File: wb/main/shared/enumerates.py [hunk 6]
```diff
@@ -14,7 +14,9 @@
  See the License for the specific language governing permissions and
  limitations under the License.
 """
+from __future__ import annotations
```
Comment (artyomtugaryov) [inline wb/main/shared/enumerates.py:17]: remove

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 7]
```diff
@@ -42,16 +46,42 @@ def run(self):
...
 
-    @staticmethod
-    def _recognize(path: str):
-        for dataset_type, dataset_recognizer in ConfigRegistry.dataset_recognizer_registry.items():
-            if dataset_recognizer.recognize(path):
-                return dataset_type
-        remove_dir(path)
-        raise UnknownDatasetError('Unknown dataset type')
+    def _recognize(self, dataset: DatasetsModel) -> DatasetTypesEnum:
+        if not DATASET_REPORTS_FOLDER.exists():
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:55]: add this code to `init_data_folders`

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 8]
```diff
@@ -42,16 +46,42 @@ def run(self):
...
+            create_empty_dir(DATASET_REPORTS_FOLDER)
+        report_path = DATASET_REPORTS_FOLDER / f'{dataset.id}.json'
+
+        tool = DatumaroTool()
+        tool.set_mode(DatumaroModesEnum.detect_format)
+        tool.set_path('json-report', report_path)
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:61]: again - create a new mwthod

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 9]
```diff
@@ -42,16 +46,42 @@ def run(self):
...
+        tool.set_path(None, dataset.path)
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:62]: the same

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: docker/dockerfiles/Dockerfile [hunk 10]
```diff
@@ -84,7 +84,9 @@ RUN apt update && \
     # For E2E in non-headless mode
     xvfb \
     gnupg2 \
-    build-essential && \
+    build-essential  \
+    # for running packages from git commits \
+    git && \
```
[the remark this answers is not in the document]
Comment (vgoluben) [reply docker/dockerfiles/Dockerfile:89]: git is required inside the docker container if we want to install python packages from specific commits

File: wb/main/shared/enumerates.py (hunk 6 above)
Comment (vgoluben) [reply wb/main/shared/enumerates.py:17]: This import is what allows for function output typing in line 63. Is it really necessary to remove this?

File: wb/main/jobs/datasets/recognize_dataset_job.py (hunk 8 above)
Comment (vgoluben) [reply wb/main/jobs/datasets/recognize_dataset_job.py:61]: We'll be getting a lot more parameters later down the line for things such as TFDS info and downloading. Both are restricted to either a flag or a path argument. Wouldn't this cause a massive function bloat inside the tool class? These generic functions were designed with this in mind.

File: wb/main/shared/enumerates.py [hunk 11]
```diff
@@ -57,5 +57,11 @@ class DatasetTypesEnum(enum.Enum):
     # text
     csv = 'csv'
 
-    def is_nlp(self):
+    @classmethod
+    def get_value(cls, value: str):
```
Comment (artyomtugaryov) [inline wb/main/shared/enumerates.py:61]: just catch the value error where you use the function

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 12]
```diff
@@ -42,16 +46,40 @@ def run(self):
...
+        with report_path.open() as report_file:
+            report = json.load(report_file)
+
+        detected_formats = report['detected_formats']
+        if not detected_formats:
+            self.cancel_with_error(dataset, 'No valid dataset format detected.')
+        elif len(detected_formats) > 1:
+            self.cancel_with_error(dataset, 'More than one valid format detected.')
+
+        dataset_type = DatasetTypesEnum.get_value(detected_formats[0])
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:75]:
```suggestion
       try:
           dataset_type = DatasetTypesEnum(detected_formats[0])
       except ValueError:
       
```

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 13]
```diff
@@ -42,16 +46,40 @@ def run(self):
...
+        if return_code:
+            self.cancel_with_error(dataset, 'Error during format detection.')
+        with report_path.open() as report_file:
+            report = json.load(report_file)
+
+        detected_formats = report['detected_formats']
+        if not detected_formats:
+            self.cancel_with_error(dataset, 'No valid dataset format detected.')
+        elif len(detected_formats) > 1:
+            self.cancel_with_error(dataset, 'More than one valid format detected.')
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:73]: just raise

File: wb/main/console_tool_wrapper/datumaro_tool/tool.py [hunk 14]
```diff
@@ -35,11 +35,13 @@ def set_conversion(self, input_format: DatasetTypesEnum, output_format: DatasetT
...
-    def set_input_output_paths(self, input_path: str, output_path: Optional[str]):
-        self.set_parameter('--input-path', input_path, ConsoleParametersTypes.path)
-        if output_path:
-            self.set_parameter('--output-dir', output_path, ConsoleParametersTypes.path)
-
-    def enable_image_copy(self):
-        self.set_parameter('--', True, ConsoleParametersTypes.flag)
-        self.set_parameter('--save-images', True, ConsoleParametersTypes.flag)
+    def set_path(self, param_name: Optional[str], path: str):
+        if path:
```
Comment (artyomtugaryov) [inline wb/main/console_tool_wrapper/datumaro_tool/tool.py:39]: remove

Comment (artyomtugaryov) [reply wb/main/console_tool_wrapper/datumaro_tool/tool.py:39]: set_images -> _set_path

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 15]
```diff
@@ -130,7 +131,7 @@ describe('UI tests on Accuracy Reports', () => {
     expect(await TestUtils.getElementByDataTestId('predictions-table').isPresent()).toBeTruthy();
   });
 
-  it('should create a project, calibrate it, create accuracy report (predictions & tensor comparisons) for an OD model', async () => {
+  fit('should create a project, calibrate it, create accuracy report (predictions & tensor comparisons) for an OD model', async () => {
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:134]: ?

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 16]
```diff
@@ -280,7 +281,7 @@ describe('UI tests on Accuracy Reports', () => {
         name: 'deeplabv3',
         framework: Frameworks.TENSORFLOW,
       };
-      await calibrationUtils.runInt8PipelineThroughDownloader(
+      /* await calibrationUtils.runInt8PipelineThroughDownloader(
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:284]: add comment with ticket number

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 17]
```diff
@@ -322,13 +323,13 @@ describe('UI tests on Accuracy Reports', () => {
...
 
-  // 73953
+  // TODO: 73953, 83042
   xit(
     'should create a project with a generic model, calibrate it, configure accuracy (OD), ' +
       'create accuracy report (predictions comparison)',
     async () => {
       const modelFile = browser.params.precommit_scope.resources.ODModels.yoloV4TinyIR;
-      await calibrationUtils.runInt8PipelineThroughUpload(
+      /* await calibrationUtils.runInt8PipelineThroughUpload(
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:332]: again

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 18]
```diff
@@ -373,15 +374,15 @@ describe('UI tests on Accuracy Reports', () => {
...
+  // TODO: 73953, 83042
   xit(
     'should create a project with a generic model and not-annotated dataset, calibrate, ' +
       'create another project with annotated dataset, ' +
       'go back to the calibrated project, configure accuracy (OD), ' +
       'go back to the second project, check that accuracy is available, run accuracy',
     async () => {
       const modelFile = browser.params.precommit_scope.resources.ODModels.ssdliteMobileNetV2;
-      await calibrationUtils.runInt8PipelineThroughUpload(
+      /* await calibrationUtils.runInt8PipelineThroughUpload(
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:385]: again

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 19]
```diff
@@ -405,7 +406,7 @@ describe('UI tests on Accuracy Reports', () => {
       await browser.sleep(1500);
 
       // Open INT8 project
-      await testUtils.homePage.openProjectByModelAndDatasetNames(modelFile.name, imageNetNotAnnotated.name, true);
+      // await testUtils.homePage.openProjectByModelAndDatasetNames(modelFile.name, imageNetNotAnnotated.name, true);
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:409]: the same

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 20]
```diff
@@ -423,7 +424,7 @@ describe('UI tests on Accuracy Reports', () => {
       expect(editorLines.includes('datasets:')).toBeTruthy();
       await browser.waitForAngularEnabled(true);
 
-      await testUtils.homePage.openProjectByModelAndDatasetNames(modelFile.name, imageNetNotAnnotated.name, true);
+      // await testUtils.homePage.openProjectByModelAndDatasetNames(modelFile.name, imageNetNotAnnotated.name, true);
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:427]: again

File: client/e2e/src/accuracy-reports.e2e-spec.ts [hunk 21]
```diff
@@ -447,14 +448,14 @@ describe('UI tests on Accuracy Reports', () => {
...
-  // 73953
+  // TODO: 73953, 83042
   xit(
     'should create a project with a classification model & ' +
       'not-annotated dataset, calibrate it, create accuracy report (predictions comparison), ' +
       'check that advanced config is filled',
     async () => {
       const modelFile = { name: 'squeezenet1.1-caffe2', framework: Frameworks.CAFFE2 };
-      await calibrationUtils.runInt8PipelineThroughDownloader(
+      /* await calibrationUtils.runInt8PipelineThroughDownloader(
```
Comment (artyomtugaryov) [inline client/e2e/src/accuracy-reports.e2e-spec.ts:458]: why is there the problem with int8?

File: client/e2e/src/int8-basic-cases.e2e-spec.ts [hunk 22]
```diff
@@ -46,41 +46,43 @@ describe('UI tests on Running Int8 Calibration', () => {
...
-  // 76569
+  // TODO: 76569, 83042
   xit(
     'Upload FP32 Mobilenet SSD Lite V2, use Not Annotated dataset, ' +
       'infer (CPU), int8 calibrate - Simplified Mode, infer (CPU)',
     async () => {
       const modelFile = browser.params.precommit_scope.resources.ODModels.ssdliteMobileNetV2;
       const inferenceTarget = InferenceType.CPU;
-      await calibrationUtils.runInt8PipelineThroughUpload(modelFile, notAnnotatedDataset, inferenceTarget);
+      // await calibrationUtils.runInt8PipelineThroughUpload(modelFile, notAnnotatedDataset, inferenceTarget);
```
Comment (artyomtugaryov) [inline client/e2e/src/int8-basic-cases.e2e-spec.ts:63]: .

File: client/e2e/src/int8-basic-cases.e2e-spec.ts [hunk 23]
```diff
@@ -46,41 +46,43 @@ describe('UI tests on Running Int8 Calibration', () => {
...
-  it(
+  // TODO: 83042
+  xit(
     'Select squeezenet1.1 from table + Not Annotated dataset, infer (CPU), int8 Simplified Mode, ' +
       'check wait message ',
     async () => {
       const modelFile = { name: 'squeezenet1.1' };
       const inferenceTarget = InferenceType.CPU;
-      const inferenceResult = await calibrationUtils.runInt8PipelineThroughDownloader(
+      /* const inferenceResult = await calibrationUtils.runInt8PipelineThroughDownloader(
```
Comment (artyomtugaryov) [inline client/e2e/src/int8-basic-cases.e2e-spec.ts:85]: .

File: wb/main/console_tool_wrapper/datumaro_tool/tool.py [hunk 24]
```diff
@@ -35,11 +35,26 @@ def set_conversion(self, input_format: DatasetTypesEnum, output_format: DatasetT
...
+        self._set_path('', path)
+
+    def _set_flag(self, param_name: str):
+        self.set_parameter(f'--{param_name}', True, ConsoleParametersTypes.flag)
+
+    def enable_image_save(self):
+        self._set_flag('save-images')
+
+    def add_separator(self):
+        self.set_parameter('--', True, ConsoleParametersTypes.flag)
```
Comment (artyomtugaryov) [inline wb/main/console_tool_wrapper/datumaro_tool/tool.py:60]: new line

File: wb/main/dataset_utils/dataset_adapters.py [hunk 25]
```diff
@@ -758,13 +758,17 @@ class VGGFace2LandmarkCustomAdapter(VGGFace2BaseAdapter):
...
-                                     for path in item.iterdir()):
+        dirs = self.dataset_path.glob('**/*')
+        for item in dirs:
+            if item.is_dir() and item.name == 'images' and all(self.is_valid_scenario_dir(path)
+                                                               for path in item.iterdir()):
                 return item
         return None
 
+    @staticmethod
+    def is_valid_scenario_dir(path: Path) -> bool:
```
Comment (artyomtugaryov) [inline wb/main/dataset_utils/dataset_adapters.py:769]: it is a bad name for the method

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: wb/main/jobs/datasets/recognize_dataset_job.py [hunk 26]
```diff
@@ -42,16 +44,41 @@ def run(self):
...
+        elif len(detected_formats) > 1:
+            raise DatumaroError('More than one valid format detected.', self.job_id)
+
+        dataset_type = None
+        try:
+            dataset_type = DatasetTypesEnum.get_value(detected_formats[0])
+        except ValueError:
+            pass
+
+        if dataset_type not in DatasetTypesEnum:
```
Comment (artyomtugaryov) [inline wb/main/jobs/datasets/recognize_dataset_job.py:79]: move to except statement

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — APPROVED** [review]

**Pull request merged**